src/coreclr/nativeaot/System.Private.CoreLib/src/System/WeakReference.NativeAot.cs

```diff
@@ -187,13 +131,26 @@ private static void TrySetComTarget(object? target)
         // Free all system resources associated with this reference.
         ~WeakReference()
         {
-#pragma warning disable 420  // FYI - ref m_handle causes this.  I asked the C# team to add in "ref volatile T" as a parameter type in a future version.
+            // Note: While WeakReference is formally a finalizable type, the finalizer does not actually run.
+            //       Instead the instances are treated specially in GC when scanning for no longer strongly-reachable
+            //       finalizable objects.
+            //
+            // It is possible that the finalizer runs for a derived type.
+
+            Debug.Assert(this.GetType() != typeof(WeakReference));
+
+            // finalizers normally do not run concurrently since finalizer queue is single-threaded.
+            // however, since this is exposed to derived types, we do not know what can happen.
             IntPtr handle = Interlocked.Exchange(ref m_handle, default(IntPtr));
-#pragma warning restore 420
             if (handle != default(IntPtr))
                 ((GCHandle)handle).Free();
+
+            // must keep the instance alive as long as we use the handle.
+            GC.KeepAlive(this);
```

Comment:
Why do we need KeepAlive in a finalizer?

Reply:
In theory a finalizer never runs concurrently with itself, thus neither `Interlocked.Exchange` nor `KeepAlive` are necessary, since noone could be recycling the handle while we are freeing it.

However, since nongeneric `WeakReference` is not sealed, it is hard to guarantee anything. 

The original code looked like it tried to handle concurrent finalization. It does look like an attempt to handle just a particular kind of misuse, so I kept such assumption.

Reply:
I guess we can remove both `Interlocked.Exchange` and `KeepAlive`. If things are broken, they can be broken in many ways.

I hope deriving from `WeakReference` or at least overriding the finalizer and doing weird things in it is not a common practice.

Reply:
It would be nice to match what CoreCLR does (work towards sharing the code eventually).

Reply:
CoreCLR has a bunch of code that handles COM and I am not very familiar with requirements. It looks a lot more complicated, perhaps because it is in native code and maybe because of the COM stuff.

Ignoring COM, there is nothing AOT specific here. WeakReferences are just thin facades to GC and handles, which are the same for CoreCLR. 
CoreCLR could be trivially switched to use managed implementation (which seems simpler), just would need to handle the COM stuff.

Reply:
The COM stuff is for http://github.com/microsoft/cswinRT/ support. If once we want to make nativeaot work well for cswinrt, we will need to implement the COM stuff too.

Reply:
> we will need to implement the COM stuff too

I'd hope there is a way to do that in managed code. 
There is some code in NativeAOT under `#if ENABLE_WINRT`, but maybe it is a different kind of WINRT.

Reply:
I looked at the native implementation a bit more. The COM is really about detecting RCWs and storing them via a different kind of handle. 

The native implementation seems to be adding a lot of complexity just for being native. There is gc protection, for the reference itself and for the referenced object. 

There is also a spinlock in the setter. Since the kind of the handle can be changing the assignment is not a single operation.

There is also a comment saying that the lock synchronizes with finalization done by GC, but GC does not run concurrently with FCalls, (not the blocking mark phases when we know which finalizables are not reachable), so I am not sure the spinlock helps with that. Besides, the getter does not take the spinlock anyways. There is comment saying it is ok, but it is not very convincing. If handle recycling is possible, it would be possible to fetch an object of a wrong type and that is still a GC hole.

I think with the eager finalization the handle recycling is not possible in `WeakReference<T>`, because eager finalization does not run concurrently with managed code. The same should hold for the nongeneric `WeakReference` as well, unless it is overriden, but then there are endless opportunities for breakage.  
Same should hold for the FCalls as well. I see some `GCX_PREEMP();` in the code, but `gc.pThis` protection should work as KeepAlive.

Maybe it is worth to actually make the native implementation closer to managed, or even switch to managed, if COM stuff can be handled via FCall callouts. Hard to tell without actually trying.